python/kserve/kserve/storage.py

```diff
@@ -212,6 +234,104 @@ def _download_gcs(uri, temp_dir: str):
+            "TLS_SKIP_VERIFY": "false",
+            "HEADERS": None,
+            "N_THREADS": "2",
+            "KERBEROS_KEYTAB": None,
+            "KERBEROS_PRINCIPAL": None,
+        }
+
+        secret_dir = _HDFS_SECRET_DIRECTORY
+        if os.environ.get("HDFS_SECRET_DIR"):
+            secret_dir = os.environ["HDFS_SECRET_DIR"]
+
+        for filename in os.listdir(secret_dir):
+            if filename not in config:
+                continue
+
+            # We don't read files which are supposed to be files, just save their path
+            if filename in _HDFS_FILE_SECRETS:
+                config[filename] = f"{secret_dir}/{filename}"
+                continue
+
+            # Read file and save value in config dict
+            with open(f"{secret_dir}/{filename}") as f:
+                config[filename] = f.read()
+
+        return config
+
+    @staticmethod
+    def _download_hdfs(uri, out_dir: str):
+        from krbcontext.context import krbContext
+        from hdfs.ext.kerberos import Client, KerberosClient
+        from hdfs.util import HdfsError
+
+        config = Storage._load_hdfs_configuration()
+
+        logging.info(f"Using the following hdfs config\n{config}")
+
+        # Remove hdfs:// from the uri to get just the path
+        # hdfs://user/me/model -> user/me/model
+        path = uri[len(_HDFS_PREFIX):]
+        if not config["HDFS_ROOTPATH"]:
+            path = "/" + path
+
+        s = requests.Session()
+
+        if config["TLS_CERT"]:
+            s.cert = (config["TLS_CERT"], config["TLS_KEY"])
+        # s.verify = , True, False, or CA PATH
+        if config["TLS_CA"]:
+            s.verify = config["TLS_CA"]
+        if config["TLS_SKIP_VERIFY"].lower() == "true":
+            s.verify = False
+
+        if config["HEADERS"]:
+            headers = json.loads(config["HEADERS"])
+            s.headers.update(headers)
+
+        if config["KERBEROS_PRINCIPAL"]:
+            context = krbContext(
+                using_keytab=True,
+                principal=config["KERBEROS_PRINCIPAL"],
+                keytab_file=config["KERBEROS_KEYTAB"]
+            )
+            context.init_with_keytab()
+            client = KerberosClient(
+                config["HDFS_NAMENODE"],
+                proxy=config["USER_PROXY"],
+                root=config["HDFS_ROOTPATH"],
+                session=s
+            )
+        else:
+            client = Client(
+                config["HDFS_NAMENODE"],
+                proxy=config["USER_PROXY"],
+                root=config["HDFS_ROOTPATH"],
+                session=s
+            )
+
+        # Check path exists
+        # Raises HdfsError when path does not exist
+        client.status(path)
+
+        try:
+            files = client.list(path)
+            for f in files:
+                client.download(f"{path}/{f}", out_dir, n_threads=int(config["N_THREADS"]))
+        except HdfsError:
+            # client.list raises exception when path is a file
```

Comment:
@markwinter Is this the only case when it throws the `HdfsError` ?

Reply:
Will double check

Reply:
It can be thrown in a few cases when calling `client.download`.

I think instead of using this exception to determine if it is a file, I will check the return of `client.status(path)` which has info on whether it is a file or directory.

Reply:
Updated